Upgrade to use multiple memory map files

MichaelKim20 opened the pull request:
The file is divided into multi parts.
The number of blocks in a file is fixed.
Store the position of each block using an array.
Replace with an index file later.

![MMFile-BlockStorage-Index](https://user-images.githubusercontent.com/11639939/65367004-5148a100-dc66-11e9-8cbe-639500e1e32c.gif)

Commit:
Improves performance when adding blocks

I had to reserve more memory than it actually was.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -0,0 +1,292 @@
+
+    public ~this ()
+    {
+        this.unMap();
```
You cannot use the GC from a destructor.
Since D has a garbage collector, when a reference is collected by the GC, you have no guarantee that the GC-allocated references it holds are still alive.
If you had this guarantee, it would be impossible to finalize objects with cyclic references.

MichaelKim20 replied:
Okay. I'll remove this part.

Commit:
Memory mapped file of a block is Implemented using a multiple file.

The file is divided into multiple parts. 
The number of blocks in a file is fixed.
Store the position of each block using an array.
 Replace with an index file later.

Commit:
Add attributes `@safe`/`@trusted`

Commit:
Create Index using `RedBlackTree`

There are two indexes.
One is the height of the block.
And the other is the hash of the block.
File I/O to the index will be added next.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -62,7 +71,40 @@ public class BlockStorage
+        ubyte[] values;
+        values.length = size_t.sizeof;
+        *cast(size_t*)&(values[0]) = length;
+        foreach (idx, ref e; values)
```
```suggestion
        foreach (idx, e; (cast(const ubyte*)&length)[0 .. size_t.sizeof])
```

Commit:
Memory map file size changes depending on data size

Commit:
Method `map` and `unMap`, change public to private

Geod24 commented:
I think it would be great to break this PR done into smaller commits.

Commit:
Store index data in the file.

When a block is stored, it is saved one by one.

Geod24 commented:
It's pretty hard to follow, as the commits are not atomic. For example the first 2 commits override each other quite a lot.
Style wise, use `private immutable ConstantName = 42;` instead of `const CONSTANT_NAME = 42;`.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
+
+    ***************************************************************************/
+
+    private void writeSizeT (size_t pos, size_t value) @trusted
```
W.r.t. `readDataLength`, `writeDataLength`, `writeSizeT`, `readSizeT`: This is what the serialization / deserialization module is for.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -37,56 +73,125 @@ public class BlockStorage
+        }
+        else
+        {
+            if (name.exists())
```
Can be turned into an `else if`

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
+        this.hash_idx.clear();
+
+        if (file_name.exists)
+        {
```
Use early returns to reduce indentation.
In this case, instead of wrapping the rest of the function in a `if`, use:
```d
if (!file_name.exists())
    return;
```

This way the code is simpler to read.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, Hash hash) @safe
```
The whole point of memory mapping is to return a `const` pointer, so that there is only ever one copy of the block. This copy the block.

Geod24 commented on source/agora/common/BlockStorage.d:
```diff
@@ -118,37 +234,60 @@ public class BlockStorage
+            (this.height_idx.back.height >= block.header.height))
             return false;
 
         const ubyte[] serialized_block = serializeFull(block);
```
Instead of allocating memory to serialize, it should be done straight to the file.

Geod24 commented:
Also can you rebase against `v0.x.x` so we get coverage report ?

Geod24 reviewed (approved):
Actually on a second take, this does not break anything, so let's move forward and fix as needed.
